### PR TITLE
sql: fix EXECUTE privileges for UDFs

### PR DESCRIPTION
**Pull request by mgartner**
#### sql: fix EXECUTE privileges for UDFs

Informs #103842

Release note (bug fix): A bug has been fixed that allowed users to
execute a user-defined function without having EXECUTE privileges on the
function. This bug has been present since version 22.2 when UDFs were
introduced.

#### upgrades: adds upgrade granting EXECUTE to public role for all functions

This commit adds an upgrade that grants EXECUTE privileges to the public
role for all existing functions. This matches the behavior of
`CREATE FUNCTION` which automatically adds this privilege as of #107317.

Release note: None

#### catpb: document `UserPrivilege.WithGrantOption`

Release note: None

### CODE REVIEW

File: pkg/sql/schema_resolver.go
Comment (mgartner) [inline pkg/sql/schema_resolver.go:602]: @chengxiong-ruan is this function expensive? We'll be calling it everytime a UDF invocation is encountered by the optimizer.

**Review by rafiss — COMMENTED** [review]
we'll need an upgrade job that does `GRANT EXECUTE ON FUNCTION f TO public` for all UDFs. it could either be done in this PR, or else it should be done separately before this PR is merged 

there can also be a logictest with the `cockroach-go-upgrade-to-master` config. it can create a UDF on the old version, upgrade the nodes, then verify that an arbitrary user can still execute the function

**Comment by mgartner** [comment]
> we'll need an upgrade job that does GRANT EXECUTE ON FUNCTION f TO public for all UDFs. it could either be done in this PR, or else it should be done separately before this PR is merged

Do you know of a PR that does something similar that I could use as a guide for this?

**Comment by rafiss** [comment]
The best recent example is probably https://github.com/cockroachdb/cockroach/pull/105832 - the upgrade step for this issue would be similar to `first_upgrade.go`.

Some others from longer ago:
- Here is a PR that modified all sequences during an upgrade: https://github.com/cockroachdb/cockroach/pull/86255/
- Here is another that modifies all tables to remove the GRANT privilege: https://github.com/cockroachdb/cockroach/pull/81310

File: pkg/sql/schema_resolver.go
Comment (chengxiong-ruan) [reply pkg/sql/schema_resolver.go:602]: it's calling `getDescriptorsByID` internally, it will hit cache of the leasing layer. Should be fine.

**Comment by mgartner** [comment]
This case seems unique from those examples in that we can't use SQL to get the fully-qualified names of each function to run `GRANT EXECUTE ON FUNCTION db.schema.func TO public` because you can't query multiple schemas from the `pg_namespace` table. Is there another method you'd recommend?

**Comment by mgartner** [comment]
Also, is this upgrade even necessary after #107317?

**Comment by chengxiong-ruan** [comment]
I think we need to read descriptors directly. There used to be a few upgrades iterating through all descriptors. For examples in 22.2, we have this one : https://github.com/cockroachdb/cockroach/blob/03d520fc55410ade5b71d284b9cbabf96085e2ef/pkg/upgrade/upgrades/descriptor_utils.go#L88-L88
With descriptor you may actually modify the privilege directly.
I think Rafi’s PR only made newly created functions benefit from the default privilege. For existing UDFs, we still need to manually modify the privileges. Fun fact that, alter default privilege only affect objects created after that…for existing tables you need to do GRANT

**Comment by mgartner** [comment]
What action is that upgrade performing? And what is the new way of performing such an upgrade - it looks like `runPostDeserializationChangesOnAllDescriptors` no longer exists?

**Comment by mgartner** [comment]
The second commit contains a first attempt at an upgrade for this, but it's not working. Is it on the right track or totally off?

**Comment by chengxiong-ruan** [comment]
> The second commit contains a first attempt at an upgrade for this, but it's not working. Is it on the right track or totally off?

It's on the right track. Nice try. Basically what you need is a way to iterate through all descriptors and update them. And you're doing the right thing.

**Comment by mgartner** [comment]
This is ready for another look.

**Review by rafiss — CHANGES_REQUESTED** [review]
i think this is close! i had testing suggestions; the suggestion for testing with a larger number of objects could also guide an improvement to the upgrade job

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107587#-:-NeJAjqh8Gzrwp7G2ENm:b-m36z6h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, @rharding6373, and @Xiang-Gu)

___
*[`pkg/upgrade/upgrades/grant_execute_to_public.go` line 61 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107587#-NeJ5zho4o7ut04FJ-Q8:-NeJ5zho4o7ut04FJ-Q9:b-7x3yhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/05a5e57c9fcb10f583b94fd6176d3d525f3e69df/pkg/upgrade/upgrades/grant_execute_to_public.go#L61)):*
> ```Go
> 	}
> 
> 	query := `SELECT id, length(descriptor) FROM system.descriptor ORDER BY id DESC`
> ```

i wonder if this SELECT should be batched, so we don't scan the entire system table all at once.

there's also a way to get them through the collection API, which might be more efficient, since it should avoid scanning all the tables and user-defined types:

```
	var fnDescs []catalog.MutableDescriptor
	if err := d.DB.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) (err error) {
		dbs, err := txn.Descriptors().GetAllDatabases(ctx, txn.KV())
		if err != nil {
			return err
		}
		return dbs.ForEachDescriptor(func(dd catalog.Descriptor) error {
			dbDesc, err := catalog.AsDatabaseDescriptor(dd)
			if err != nil {
				return err
			}
			schemas, err := txn.Descriptors().GetAllSchemasInDatabase(ctx, txn.KV(), dbDesc)
			if err != nil {
				return err
			}
			return schemas.ForEachDescriptor(func(sd catalog.Descriptor) error {
				scDesc, err := catalog.AsSchemaDescriptor(dd)
				if err != nil {
					return err
				}
				return scDesc.ForEachFunctionSignature(func(sig descpb.SchemaDescriptor_FunctionSignature) error {
					fnDesc, err := txn.Descriptors().MutableByID(txn.KV()).Desc(ctx, sig.ID)
					if err != nil {
						return err
					}
					fnDescs = append(fnDescs, fnDesc)
					return nil
				})
			})
		})
	}); err != nil {
		return err
	}
```

this is roughly the way that vtables like pg_catalog.pg_proc are populated

i think that way might be better actually. as-is, this upgrade is modifying the descriptors concurrently with scanning through the system.descriptors table. with the above suggestion, it will fetch the descriptors first, and then you can update them in batches.

whatever the approach, it would be worth adding a test/bench case to grant_execute_to_public_test.go that runs with a large number of objects in the database.
___
*[`pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges` line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107587#-NeJ410-CI3N5hoSn8cx:-NeJ410-CI3N5hoSn8cy:b-owoa7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f91bf214187ab5ce2ef375912ddf6f21e78677b/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges#L70)):*
> ```
> SELECT crdb_internal.node_executable_version() SIMILAR TO '23.1-%'
> ----
> true
> ```

i think this test could have an assertion like this:
```
# Makes sure the upgrade job has finished, and the cluster version gate is passed. (1000000 is the offset for dev cluster versions)
query B retry  
SELECT crdb_internal.is_at_least_version('1000023.1-24')
----  
true

# Makes sure the user still has the ability to execute f.
query I
SELECT f()
----
1
```

**Review by rharding6373 — APPROVED** [review]
:lgtm: pending Rafi's suggestions. Thanks for fixing this!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107587#-:-NeJa2VW9pP_nwX44L8O:b3a6dt3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @Xiang-Gu)

**Comment by mgartner** [comment]
*[`pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges` line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107587#-NeJ410-CI3N5hoSn8cx:-NeOQWiw9XHkTcEmbhXT:b-pb9t9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f91bf214187ab5ce2ef375912ddf6f21e78677b/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges#L70)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think this test could have an assertion like this:
```
# Makes sure the upgrade job has finished, and the cluster version gate is passed. (1000000 is the offset for dev cluster versions)
query B retry  
SELECT crdb_internal.is_at_least_version('1000023.1-24')
----  
true

# Makes sure the user still has the ability to execute f.
query I
SELECT f()
----
1
```
</blockquote></details>

Is there a way to wait for the upgrade job to finish, besides sleeping which is prone to flakes? From my conversations with @Xiang-Gu I was under the impression that this test is only to test the mixed-version state, not the upgrade. This is why I wrote another test in `grant_execute_to_public_test.go`. Is that incorrect?

**Comment by mgartner** [comment]
*[`pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges` line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107587#-NeJ410-CI3N5hoSn8cx:-NeOQzkND-LTixErDCXh:b4g6xxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f91bf214187ab5ce2ef375912ddf6f21e78677b/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges#L70)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Is there a way to wait for the upgrade job to finish, besides sleeping which is prone to flakes? From my conversations with @Xiang-Gu I was under the impression that this test is only to test the mixed-version state, not the upgrade. This is why I wrote another test in `grant_execute_to_public_test.go`. Is that incorrect?
</blockquote></details>

Oh, I see the retry in the query above. I'll add that. Should I delete the test in `gran_execute_to_public_test.go` then?

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107587)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @rafiss, and @Xiang-Gu)

___
*[`pkg/upgrade/upgrades/grant_execute_to_public.go` line 61 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/107587#-NeJ5zho4o7ut04FJ-Q8:-NeP7nk58HjNvqr7Xxwf:bzeo0bi) ([raw file](https://github.com/cockroachdb/cockroach/blob/05a5e57c9fcb10f583b94fd6176d3d525f3e69df/pkg/upgrade/upgrades/grant_execute_to_public.go#L61)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i wonder if this SELECT should be batched, so we don't scan the entire system table all at once.

there's also a way to get them through the collection API, which might be more efficient, since it should avoid scanning all the tables and user-defined types:

```
	var fnDescs []catalog.MutableDescriptor
	if err := d.DB.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) (err error) {
		dbs, err := txn.Descriptors().GetAllDatabases(ctx, txn.KV())
		if err != nil {
			return err
		}
		return dbs.ForEachDescriptor(func(dd catalog.Descriptor) error {
			dbDesc, err := catalog.AsDatabaseDescriptor(dd)
			if err != nil {
				return err
			}
			schemas, err := txn.Descriptors().GetAllSchemasInDatabase(ctx, txn.KV(), dbDesc)
			if err != nil {
				return err
			}
			return schemas.ForEachDescriptor(func(sd catalog.Descriptor) error {
				scDesc, err := catalog.AsSchemaDescriptor(dd)
				if err != nil {
					return err
				}
				return scDesc.ForEachFunctionSignature(func(sig descpb.SchemaDescriptor_FunctionSignature) error {
					fnDesc, err := txn.Descriptors().MutableByID(txn.KV()).Desc(ctx, sig.ID)
					if err != nil {
						return err
					}
					fnDescs = append(fnDescs, fnDesc)
					return nil
				})
			})
		})
	}); err != nil {
		return err
	}
```

this is roughly the way that vtables like pg_catalog.pg_proc are populated

i think that way might be better actually. as-is, this upgrade is modifying the descriptors concurrently with scanning through the system.descriptors table. with the above suggestion, it will fetch the descriptors first, and then you can update them in batches.

whatever the approach, it would be worth adding a test/bench case to grant_execute_to_public_test.go that runs with a large number of objects in the database.
</blockquote></details>

Done. The added benchmark showed no difference in performance between the two, but the code is much cleaner now I think.
___
*[`pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges` line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107587#-NeJ410-CI3N5hoSn8cx:-NeP7zXvBhF4hWgJFfA7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f91bf214187ab5ce2ef375912ddf6f21e78677b/pkg/sql/logictest/testdata/logic_test/mixed_version_udf_execute_privileges#L70)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Oh, I see the retry in the query above. I'll add that. Should I delete the test in `gran_execute_to_public_test.go` then?
</blockquote></details>

Done.

**Review by rafiss — APPROVED** [review]
thanks for the work on this!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107587#-:-NeiXbNS-tMmVR5sfm-z:br91krd)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @Xiang-Gu)

**Comment by mgartner** [comment]
TFTR!

bors r+

**Pull request merged**
